nntrainer/layers/common_properties.cpp

```diff
@@ -58,6 +59,86 @@ bool InputSpec::isValid(const ConnectionSpec &v) const {
   return v.getLayerIds().size() > 0;
 }
 
+/**
+ * @brief unsigned integer property, internally used to parse padding values
+ *
+ */
+class Padding_ : public nntrainer::Property<int> {
+public:
+  using prop_tag = int_prop_tag; /**< property type */
+};
+
+bool Padding2D::isValid(const std::string &v) const {
+
+  /// case 1, 2: padding has string literal
+  if (istrequal(v, "valid") || istrequal(v, "same")) {
+    return true;
+  }
+
+  std::vector<props::Padding_> paddings;
+  from_string(v, paddings);
+
+  /// case 3, 4, 5: padding has a sequence of unsigned integer
+  if (paddings.size() == 1 || paddings.size() == 2 || paddings.size() == 4) {
+    /// check if every padding is non-negative integer
+    for (const auto &padding : paddings) {
+      if (padding.get() < 0) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /// case else: false
+  return false;
+}
+
+std::vector<unsigned int> Padding2D::compute(const TensorDim &input,
+                                             const TensorDim &kernel) {
+  auto &padding_repr = get(); /// padding representation
+  if (istrequal(padding_repr, "valid")) {
+    return {0, 0, 0, 0};
+  }
+
+  if (istrequal(padding_repr, "same")) {
+    /// @note if we start to consider dilation, this calculation has to tuned
+    /// accordingly.
+    auto calculate_padding = [](unsigned input_, unsigned kernel_) {
+      NNTR_THROW_IF(input_ < kernel_, std::invalid_argument)
+        << "input smaller then kernel not supported, input size: " << input_
+        << " kernel size: " << kernel_ << " padding: same\n";
+      return kernel_ - 1;
+    };
+
+    auto pad_horizontal = calculate_padding(input.width(), kernel.width());
+    auto pad_vertical = calculate_padding(input.height(), kernel.height());
+
+    auto pad_top = pad_vertical / 2;
+    auto pad_left = pad_horizontal / 2;
+
+    return {pad_top, pad_vertical - pad_top, pad_left,
```

Comment:
maybe these can be cached? it depends on if the compute function is to be called multiple times. If it is called once/twice, its fine.

Reply:
It is supposed to called at the finalize only, so I think it's fine :)

Reply:
then its alright.